### PR TITLE
Fix #1830

### LINKED ISSUE #1830: [REQUEST] Suggest use of NVCC's --keep-device-functions switch
**Is your feature request related to a problem? Please describe.**
The "problem" is that if you compile device-side functions with no kernels - NVCC won't generate any PTX code. So you end up having to write kernels

**Describe the solution you'd like**
1. Best option: A button for keeping vs discarding device functions.
2. Second best: A suggestion of this option in the box which opens up with compiler options when you click the button on the side of the compiler switches line.

I'm of two minds regarding whether or not the button should be on by default or not.

**Describe alternatives you've considered**
Well, you can enter the NVCC switch yourself - if you know about it. Most people don't.

### PR DESCRIPTION
**Pull request by guoci**
For #1830

### CODE REVIEW

**Comment by mattgodbolt** [comment]
Can we have a little more explanation as to what's going on here please?

File: lib/compiler-finder.ts [hunk 1]
```diff
@@ -203,6 +203,10 @@ export class CompilerFinder {
 
         const props: PropertyGetter = (propName: string, defaultValue?: any) => {
             const propsForCompiler = parentProps(langId, base + propName);
+            if (propsForCompiler && propName === 'options') {
```
Comment (mattgodbolt) [inline lib/compiler-finder.ts:206]: This seems like a very arbitrary and hardcoded, but also very big change to how properties work. I'm not in favour of this kind of thing: there shouldn't be special cases in the generic property system for a particularly-named option.

**Review by mattgodbolt — CHANGES_REQUESTED** [review]
I'm not sure quite why this is needed, and why tihs fixes #1830. I'm not comfortable with a special case in this general purpose bit of code. If we really do need cascading properties we should support that natively somehow and not unconditionally append them.

File: lib/compiler-finder.ts (hunk 1 above)
Comment (guoci) [reply lib/compiler-finder.ts]: Allow cascading properties for all `propName` was breaking the application, so the special case for `options` was added. But I have pushed a new commit to remove the special case and fix the problem in another way.

**Comment by guoci** [comment]
In `cuda.amazon.properties`, there are 2 lines where the compiler options are set:
```
group.nvcc.options=--keep-device-functions
compiler.nvcc121.options=--compiler-bindir /opt/compiler-explorer/gcc-10.2.0/bin
```
When the `props`gets the `options` property, only `compiler.nvcc121.options` will be used, so `--keep-device-functions` will not be in the options list.

**Comment by partouf** [comment]
Ooohhhh riiight now I get it... But this might/will actually break other things, the properties were never meant to concatenate, always to overwrite.

Good thinking though and noticing the option was actually there, I had totally missed that... But I think this should be solved by adding `--keep-device-functions` to the `nvcc.ts` file instead as to not break anything else.

**Comment by guoci** [comment]
@partouf I have pushed the change as you described.

**Comment by partouf** [comment]
> @partouf I have pushed the change as you described.

Thanks!

**Comment by guoci** [comment]
Is this live yet? Doesn't seem to work.

**Comment by partouf** [comment]
no this is not live yet

**Comment by mattgodbolt** [comment]
It is now live :) Thanks for your patience.

**Pull request merged**